kotlinx-coroutines-core/native/src/MultithreadedDispatchers.kt

```diff
@@ -67,28 +68,81 @@ internal class WorkerDispatcher(name: String) : CloseableCoroutineDispatcher(),
     }
 }
 
-private class MultiWorkerDispatcher(name: String, workersCount: Int) : CloseableCoroutineDispatcher() {
+private const val IS_CLOSED_MASK = 1 shl 31
+
+private class MultiWorkerDispatcher(
+    private val name: String,
+    private val workersCount: Int
+) : CloseableCoroutineDispatcher() {
     private val tasksQueue = Channel<Runnable>(Channel.UNLIMITED)
-    private val workers = Array(workersCount) { Worker.start(name = "$name-$it") }
+    private val workers = atomicArrayOfNulls<Worker>(workersCount)
+
+    // Number of active workers + isClosed flag in the highest bit
+    private val controlState = atomic(0)
+    private val activeWorkers: Int get() = controlState.value and (IS_CLOSED_MASK.inv())
+    private var isClosed: Boolean
+        get() = controlState.value.isClosed()
+        set(value) {
+            assert { value }
+            controlState.value = controlState.value or IS_CLOSED_MASK
```

Comment:
We may still leak some workers this way.
1. `controlState.value` is read in the `isClosed = true` statement.
2. Some new workers arrive. They increment `controlState.value` and do nothing else.
3. The old `controlState.value` is `or`-ed with `IS_CLOSED_MASK` and written to `controlState.value`.
4. The cleanup procedure is done.
5. The workers add themselves to the array and begin to do their thing.

Reply:
Good catch, thanks